File: src/Core.Abstractions/SpecializationCache.cs
namespace Zeta.Core;

/// <summary>
/// WeakRef-wrapped specialization cache — cogen=mix(mix,mix) as memory management.
/// The specialized delegate is weakly held. If GC collects it, the next call
/// regenerates from the IR. NEVER caches errors.
/// </summary>
/// <typeparam name="TInput">Input type.</typeparam>
/// <typeparam name="TOutput">Output type.</typeparam>
public sealed class SpecializationCache<TInput, TOutput>
{
    private readonly Func<Func<TInput, TOutput>> _specializer;
    private WeakReference<Holder>? _cached;
    private int _hits;
    private int _misses;
    private int _errors;

    private sealed class Holder
    {
        public required Func<TInput, TOutput> Fn { get; init; }
    }

    public SpecializationCache(Func<Func<TInput, TOutput>> specializer)
    {
        _specializer = specializer;
    }

    public int Hits => _hits;
    public int Misses => _misses;
    public int Errors => _errors;

    /// <summary>Run the specialized function. Specializes on first call, uses cache after.</summary>
    public TOutput Run(TInput input)
    {
        var fn = GetOrRegenerate();
        return fn(input);
    }

    /// <summary>Force regeneration on next call (invalidate cache).</summary>
    public void Invalidate() => _cached = null;

    private Func<TInput, TOutput> GetOrRegenerate()
    {
        if (_cached is not null && _cached.TryGetTarget(out var holder))
        {
            Interlocked.Increment(ref _hits);
            return holder.Fn;
        }

        Interlocked.Increment(ref _misses);
        try
        {
            var fn = _specializer();
            var newHolder = new Holder { Fn = fn };
            _cached = new WeakReference<Holder>(newHolder);
            return fn;
        }
        catch
        {
            // NEVER cache errors — always retry on next call
            Interlocked.Increment(ref _errors);
            _cached = null;
            throw;
        }
    }
}

File: src/Core/SpecializationCache.fs
namespace Zeta.Core

open System

/// WeakRef-wrapped specialization cache — cogen=mix(mix,mix) as memory management.
/// The specialized function is weakly held. If GC collects it, next call regenerates.
/// NEVER caches errors (always retries on next call).
[<Sealed>]
type SpecializationCache<'TInput, 'TOutput>(specializer: unit -> ('TInput -> 'TOutput)) =
    let mutable cached: WeakReference<'TInput -> 'TOutput> option = None
    let mutable hits = 0
    let mutable misses = 0
    let mutable errors = 0

    /// Number of cache hits (specialized function was still alive).
    member _.Hits = hits

    /// Number of cache misses (first call or GC collected it).
    member _.Misses = misses

    /// Number of specialization errors (never cached).
    member _.Errors = errors

    /// Run the specialized function. Specializes on first call, uses cache after.
    member this.Run(input: 'TInput) : 'TOutput =
        let fn = this.GetOrRegenerate()
        fn input

    /// Force regeneration on next call.
    member _.Invalidate() = cached <- None

    member private _.GetOrRegenerate() : ('TInput -> 'TOutput) =
        match cached with
        | Some wr ->
            match wr.TryGetTarget() with
            | true, fn ->
                hits <- hits + 1
                fn
            | _ ->
                // GC collected it — fall through to regenerate
                misses <- misses + 1
                SpecializationCache.Regenerate(&cached, &errors, specializer)
        | None ->
            misses <- misses + 1
            SpecializationCache.Regenerate(&cached, &errors, specializer)

    static member private Regenerate
        (cached: byref<WeakReference<'TInput -> 'TOutput> option>,
         errors: byref<int>,
         specializer: unit -> ('TInput -> 'TOutput)) =
        try
            let fn = specializer()
            cached <- Some(WeakReference<_>(fn))
            fn
        with _ ->
            // NEVER cache errors
            errors <- errors + 1
            cached <- None
            reraise()
